LoadTests.cmake

```diff
@@ -0,0 +1,47 @@
+# This file is include()'d by CTest. It executes test-libmongoc to get a list
+# of all tests that are registered. Each test is then defined as a CTest test,
+# allowing CTest to control the execution, parallelization, and collection of
+# test results.
+
+if (NOT EXISTS "${TEST_LIBMONGOC_EXE}")
+    # This will fail if 'test-libmongoc' is not compiled yet.
+    message (WARNING "The test executable ${TEST_LIBMONGOC_EXE} is not present. "
+                     "Its tests will not be registered")
+    return ()
+endif ()
+
+# Get the list of tests
+execute_process (
+    COMMAND "${TEST_LIBMONGOC_EXE}" --list-tests --no-fork
+    OUTPUT_VARIABLE tests_out
+    WORKING_DIRECTORY "${SRC_ROOT}"
+    RESULT_VARIABLE retc
+    )
+if (retc)
+    # Failed to list the tests. That's bad.
+    message (FATAL_ERROR "Failed to run test-libmongoc to discover tests [${retc}]:\n${tests_out}")
+endif ()
+
+# Split lines on newlines
+string (REPLACE "\n" ";" lines "${tests_out}")
+
+# Generate the test definitions
+foreach (line IN LISTS lines)
+    if (NOT line MATCHES "^/")
+        # Only generate if the line begins with `/`, which all tests should.
+        continue ()
+    endif ()
+    # The new test name is prefixed with 'mongoc'
+    set (test "mongoc/${CONFIG}${line}")
```

Comment:
Qualifying them by the configuration is required in the case of multiconf builds, in which case tests can be registered for each build type simultaneously. CTest with multiconf builds is a bit goofy. Multiconf builds are goofy in general.

Reply:
You made me curious, as VS's `RUN_TESTS` "target" isn't something I have inspected before. It looks to be that `RUN_TEST` simply runs CTest with `--build-config $(Configuration)`, which will match what was specified in VS or MSBuild. If you try to run `ctest` in a VS-generated build directory without specifying the build type, CTest will not attempt to define any of the tests that are based on multi-conf generated executables, and will instead generate invalid test definitions. I've updated the code to mimic this behavior with multi-conf generators and require a `-C` to be specified. This will thus only register each test once, and we can omit the build-type qualifier from the names. (Previously it would try to register a test for each available config, thus needing the qualified names).

Aside: The `-R` argument takes a regex, not a glob, and it searches for a submatch. `-R '/bson/*` only just happens to also match the same tests 🙂. I actually find this a bit easier since I can just write a sub-string of the test name to `-R`, no quoting required.

---

Not really relevant here, but based on how Catch2 does it, I hypothesize that it actually contains a subtle bug in an uncommon situation where you build multiple configurations simultaneously (which made possible with the `Ninja Multi-Config` generator). The test registration runs as a post-build step on each target, and writes those tests to a file. The file it generates is not qualified by the build type, so whichever configuration was the last to build will end up writing the test registration file. In most situations this won't matter, but in a simultaneous multi-conf build each configuration will end up racing to write the test listing, and the last one to execute will end up being the one that it registered.

But that's just a theory: a _CTest_ theory.